File: app/src/main/java/com/mmjang/ankihelper/data/database/DatabaseContext.java
package com.mmjang.ankihelper.data.database;

import android.content.Context;
import android.content.ContextWrapper;
import android.content.SharedPreferences;
import android.database.DatabaseErrorHandler;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteDatabase.CursorFactory;
import android.util.Log;

import com.mmjang.ankihelper.util.StorageManager;

import java.io.File;

public class DatabaseContext extends ContextWrapper {

    private static final String DEBUG_CONTEXT = "DatabaseContext";
    private final StorageManager storageManager;

    public DatabaseContext(Context base) {
        super(base);
        SharedPreferences preferences = base.getSharedPreferences("ankihelper_prefs", Context.MODE_PRIVATE);
        this.storageManager = new StorageManager(base, preferences);
    }

    @Override
    public File getDatabasePath(String name)  {
        // Always use external storage - no fallback to internal storage
        File databaseDir = storageManager.getDatabaseDir();
        if (databaseDir == null) {
            Log.e(DEBUG_CONTEXT, "Database directory is null!");
            return null;
        }

        // Don't modify the database name, just use it as-is
        File result = new File(databaseDir, name);

        if (!result.getParentFile().exists()) {
            result.getParentFile().mkdirs();
        }

        return result;
    }

    /* this version is called for android devices >= api-11. thank to @damccull for fixing this. */
    @Override
    public SQLiteDatabase openOrCreateDatabase(String name, int mode, SQLiteDatabase.CursorFactory factory, DatabaseErrorHandler errorHandler) {
        return openOrCreateDatabase(name,mode, factory);
    }

    /* this version is called for android devices < api-11 */
    @Override
    public SQLiteDatabase openOrCreateDatabase(String name, int mode, SQLiteDatabase.CursorFactory factory) {
        File dbPath = getDatabasePath(name);
        SQLiteDatabase result = SQLiteDatabase.openOrCreateDatabase(dbPath.getAbsolutePath(), null);

        Log.w(DEBUG_CONTEXT, "openOrCreateDatabase(" + name + ",,) = " + result.getPath());
        return result;
    }
}

File: app/src/main/java/com/lmyby/ankiquicker/data/database/DatabaseContext.kt
package com.lmyby.ankiquicker.data.database

import android.content.Context
import android.content.ContextWrapper
import android.database.DatabaseErrorHandler
import android.database.sqlite.SQLiteDatabase
import android.util.Log
import com.lmyby.ankiquicker.util.StorageManager
import java.io.File

/**
 * Converted to Kotlin as part of Phase 2 data model migration
 */
class DatabaseContext(base: Context) : ContextWrapper(base) {

    private val storageManager: StorageManager

    init {
        val preferences = base.getSharedPreferences("ankihelper_prefs", Context.MODE_PRIVATE)
        this.storageManager = StorageManager(base, preferences)
    }

    override fun getDatabasePath(name: String): File? {
        // Always use external storage - no fallback to internal storage
        val databaseDir = storageManager.getDatabaseDir() ?: run {
            Log.e(DEBUG_CONTEXT, "Database directory is null!")
            return null
        }

        // Don't modify the database name, just use it as-is
        val result = File(databaseDir, name)

        if (!result.parentFile!!.exists()) {
            result.parentFile!!.mkdirs()
        }

        return result
    }

    /* this version is called for android devices >= api-11. thank to @damccull for fixing this. */
    override fun openOrCreateDatabase(
        name: String,
        mode: Int,
        factory: SQLiteDatabase.CursorFactory?,
        errorHandler: DatabaseErrorHandler?
    ): SQLiteDatabase {
        return openOrCreateDatabase(name, mode, factory)
    }

    /* this version is called for android devices < api-11 */
    override fun openOrCreateDatabase(
        name: String,
        mode: Int,
        factory: SQLiteDatabase.CursorFactory?
    ): SQLiteDatabase {
        val dbPath = getDatabasePath(name)
        val result = SQLiteDatabase.openOrCreateDatabase(dbPath!!.absolutePath, null)

        Log.w(DEBUG_CONTEXT, "openOrCreateDatabase($name,,) = ${result.path}")
        return result
    }

    companion object {
        private const val DEBUG_CONTEXT = "DatabaseContext"
    }
}
